bundles/org.eclipse.swt/Eclipse SWT/win32/org/eclipse/swt/widgets/Control.java

```diff
@@ -1071,6 +1071,7 @@ public boolean forceFocus () {
 	Decorations shell = menuShell ();
 	shell.setSavedFocus (this);
 	if (!isEnabled () || !isVisible () || !isActive ()) return false;
+	if (display.getActiveShell() != shell && !Display.isActivateShellOnForceFocus()) return false;
```

Comment:
Regarding `display.getActiveShell() != shell` :
The condition looks weird in this code. If the Shell is already active, does it make sense to activate it again? I would say that no. I agree that previously it already worked like that, but adding another confusing piece to this code makes it too many already.

The other problem is the `return false` part you're adding.
This function's logic is somewhat hard to grasp (as in, one has to stop and think to fully understand it), and that early return only adds to confusion, because it's not exactly an error and is more like an expected behavior. If you don't agree yet, try to answer this question quickly: should this line be moved up or down or stay exactly where it is?

After meditating on this code for a bit, I think I would prefer it to be laid out this way:
```java
boolean activateShellOnFocus(Shell shell) {
	// Already active?
	if (display.getActiveShell() == shell) return true;

	// Issue 450: When forcing focus to a control owned by a background Shell, the Shell should stay in background
	if (!Display.isActivateShellOnForceFocus()) return false;

	// <Here, place code to activate Shell from `Control.forceFocus()`)
}

public boolean forceFocus () {
	// <function prologue is kept as is>

	// Only focus controls in active shells. Otherwise, `shell.setSavedFocus()` will cause Control to be activated the next time Shell becomes active.
	if (!activateShellOnFocus(shell)) return false;

	// <Here, keep code that was focusing Control after activating Shell>
}
```

Note that `return true` on first line of `activateShellOnFocus()` modifies your patch and needs to be tested. But it makes things easier to understand and hopefully removes unneeded API calls. Otherwise, you could leave it as it was, but explain in the code comment that you didn't risk to change old behavior even if it sounded reasonable.

Reply:
I think, changing the code - like you suggest - can be done later if there is time or a good reason to do it, e.g. fixing another problem. At the moment, I think, your suggested much larger changes would just delay this fix for at least another couple of months and comes with a higher risk of breaking something else.

Reply:
In such case, please still extract the function, but keep code logically the same and add an explaining code comment. Adding more complexity to already complex code is not a good thing to do.

Reply:
I'd disagree. My intention for this whole RFE is to separate setting focus from activating the shell. The `Display.isActivateShellOnForceFocus()` is only a safety net for users who report a regression by this RFE to test-drive the current behavior and have a quick fix. In a second step (1 year later?) I would go ahead and remove `Display.isActivateShellOnForceFocus()` if no regression would have been reported. I think, an extra method `activateShellOnFocus()` would carve the existing behavior in stone, contradicting the separation of tasks.

Reply:
I anticipate that code won't be removed (simply because everyone would forget to). Also, deleting a dedicated function is easier than figuring which lines of code to delete. Either way, even though I don't like your approach, I won't block the patch over it. Consider this a suggestion.